File: worker/battle_cloud_worker/profiles.py
"""Profiles, mirroring `api/src/contract/profiles.ts`.

The API's camelCase field names and `analyze_replay`'s snake_case keyword arguments meet
here and nowhere else, which is why the mapping is written out explicitly rather than
derived by transforming names.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Mapping


@dataclass(frozen=True)
class EngineParams:
    """One profile's expansion.

    `search_budget_ms_per_turn` is the whole per-turn budget. The engine divides it across
    samples itself (`per_sample_ms = max(1, search_time_ms // n_opponent_samples)`), so it
    must be passed through as-is rather than pre-divided here.
    """

    search_budget_ms_per_turn: int
    opponent_samples: int
    threads: int
    usage_stats_cutoff: int
    seed: int

    def as_analyze_kwargs(self) -> Dict[str, Any]:
        """The exact keyword arguments `battle_engine.replay_analysis.analyze_replay`
        declares. Names are pinned to that signature; changing one silently passes an
        unexpected keyword and raises a `TypeError` at call time."""
        return {
            "search_time_ms": self.search_budget_ms_per_turn,
            "n_opponent_samples": self.opponent_samples,
            "threads": self.threads,
            "usage_stats_cutoff": self.usage_stats_cutoff,
            "seed": self.seed,
        }


PROFILES: Mapping[str, EngineParams] = {
    "ladder-parity": EngineParams(1000, 8, 4, 1500, 0),
    "quick": EngineParams(200, 2, 4, 1500, 0),
}

File: api/src/contract/profiles.ts
/**
 * Engine parameters, keyed by profile name.
 *
 * A client names a profile and never sends raw search parameters. The reason is not
 * ergonomics: `search_time_ms` is a CPU budget on this server, and letting a caller pick
 * it hands them a lever on machine load. See docs/contract.md.
 *
 * Field names here are the API's camelCase spelling. The worker maps them to
 * `analyze_replay`'s snake_case keyword arguments, which is the only place the two
 * spellings meet.
 */
export const PROFILES = {
  'ladder-parity': {
    searchBudgetMsPerTurn: 1000,
    opponentSamples: 8,
    threads: 4,
    usageStatsCutoff: 1500,
    seed: 0,
  },
  quick: {
    searchBudgetMsPerTurn: 200,
    opponentSamples: 2,
    threads: 4,
    usageStatsCutoff: 1500,
    seed: 0,
  },
} as const;

export type Profile = keyof typeof PROFILES;

export type EngineParams = (typeof PROFILES)[Profile];

export const PROFILE_NAMES = Object.keys(PROFILES) as [Profile, ...Profile[]];

export function isProfile(value: unknown): value is Profile {
  return typeof value === 'string' && Object.hasOwn(PROFILES, value);
}

/**
 * Wall-clock search time a replay of `turns` turns will cost under `profile`, in
 * milliseconds. `searchBudgetMsPerTurn` is the whole per-turn budget, already divided
 * across opponent samples by the engine, so this is turns times budget and not turns
 * times budget times samples.
 *
 * Search only. Parsing, state translation, and process startup are not included, so
 * this is a floor rather than an estimate of total job duration.
 */
export function searchTimeMs(turns: number, profile: Profile): number {
  return turns * PROFILES[profile].searchBudgetMsPerTurn;
}
